### PR TITLE
feat: Phase 4 — comparative plots (block_plot, youden_plot, star_plot)

### PR DESCRIPTION
**Pull request by Pjieter**
## Summary

- Adds `src/drippy/comparative.py` with 3 new NIST EDA plots (Phase 4, the final phase):
  - `block_plot` (NIST 1.3.3.3): treatment vs block effects in a blocking design
  - `youden_plot` (NIST 1.3.3.31): interlab comparison with equality line, median reference lines, and optional `doe` variant
  - `star_plot` (NIST 1.3.3.29): multivariate radar/star chart with per-variable 0–1 normalization
- Adds `tests/test_comparative.py` with 26 tests covering return types, validation, custom fig/ax, visual properties, and fluent method delegation
- Updates `src/drippy/__init__.py` to export the 3 new functions (37 total public names)
- No changes to `data.py` — fluent method stubs were already in place

## Test plan

- [x] `uv run pytest tests/test_comparative.py -v` — 26 new tests pass
- [x] `uv run pytest tests/ -v` — full suite passes (236 tests, 0 failures)
- [x] `uv run ruff check src/drippy/comparative.py tests/test_comparative.py src/drippy/__init__.py` — clean

https://claude.ai/code/session_014yUeoDbS4TxRj7vXyhGyvL

---
_Generated by [Claude Code](https://claude.ai/code/session_014yUeoDbS4TxRj7vXyhGyvL)_

## Summary by CodeRabbit

* **New Features**
  * Three new comparative visualizations are available: block plots (treatment/block comparisons), Youden plots (with reference lines and optional DOE markers), and star (radar) plots for multivariate exploration. These are exported on the package top-level for direct import.

* **Tests**
  * Added comprehensive tests covering visuals, labels/legends, error conditions, and fluent plotting methods.

[Review Change Stack](https://app.coderabbit.ai/change-stack/Pjieter/drippy/pull/24)

### CODE REVIEW

File: src/drippy/comparative.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    Raises:
+        ValueError: If factors is None or missing required keys.
+    """
+    if (
+        not data.factors
+        or "treatment" not in data.factors
+        or ("block" not in data.factors)
+    ):
+        msg = "block_plot requires factors with 'treatment' and 'block' keys"
+        raise ValueError(msg)
```
[the remark this answers is not in the document]
Comment (Pjieter) [reply src/drippy/comparative.py:45]: @coderabbitai `EDAData.__init__` already validates all inputs at construction time — it checks that `y` is non-empty and 1-dimensional, and that every entry in `factors` matches `len(y)`. Any empty array or length mismatch raises `ValueError` before `block_plot` is ever called. Adding the same validation here would be unreachable dead code.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/drippy/comparative.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    ax.legend(title="Block")
+    fig.tight_layout()
+    return fig, ax
+
+
+def youden_plot(
+    data: EDAData,
+    fig: Figure | None = None,
+    ax: Axes | None = None,
+    doe: bool = False,
```
[the remark this answers is not in the document]
Comment (Pjieter) [reply src/drippy/comparative.py:74]: @coderabbitai `FBT` (boolean-trap) rules are explicitly in the project's ruff ignore list in `pyproject.toml`. The `doe: bool = False` parameter is also the same pattern already used in `contour_plot` (multifactor.py), so this is consistent with the established codebase convention.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/drippy/comparative.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+    Returns:
+        The figure and axes containing the plot.
+
+    Raises:
+        ValueError: If x is None.
+    """
+    if data.x is None:
+        msg = "youden_plot requires x"
+        raise ValueError(msg)
```
[the remark this answers is not in the document]
Comment (Pjieter) [reply src/drippy/comparative.py:96]: @coderabbitai `EDAData.__init__` validates all inputs at construction time — `y` is required to be non-empty and 1D, and `x` must match `len(y)` if provided. Any empty array or length mismatch raises `ValueError` before `youden_plot` is ever called. Adding duplicate validation here would be unreachable dead code.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/drippy/comparative.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+    Returns:
+        The figure and axes containing the plot.
+
+    Raises:
+        ValueError: If factors is None.
+    """
+    if not data.factors:
+        msg = "star_plot requires factors"
+        raise ValueError(msg)
```
[the remark this answers is not in the document]
Comment (Pjieter) [reply src/drippy/comparative.py:142]: @coderabbitai `EDAData.__init__` validates all inputs at construction time — `y` is required to be non-empty and 1D, and every entry in `factors` must match `len(y)`. Any empty array or length mismatch raises `ValueError` before `star_plot` is ever called. Adding duplicate validation here would be unreachable dead code.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/drippy/comparative.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    if fig is None and ax is None:
+        fig, ax = plt.subplots(subplot_kw={"projection": "polar"})
+    elif fig is not None and ax is None:
+        ax = fig.add_subplot(1, 1, 1, projection="polar")
+    elif fig is None:
+        fig = ax.get_figure()
```
[the remark this answers is not in the document]
Comment (Pjieter) [reply src/drippy/comparative.py:149]: @coderabbitai `get_figure_and_axes` creates a standard `Axes` with no projection support — it cannot be used here. `star_plot` requires a `PolarAxes`, which can only be obtained via `plt.subplots(subplot_kw={"projection": "polar"})`. The same custom-creation pattern is used in `bihistogram` (onefactor.py), which also bypasses `get_figure_and_axes` because it needs a specific subplot layout.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/drippy/comparative.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+    var_names = ["y", *list(data.factors.keys())]
+    factor_cols = [data.factors[k] for k in data.factors]
+    all_vals = np.column_stack([data.y, *factor_cols])
```
[the remark this answers is not in the document]
Comment (Pjieter) [reply src/drippy/comparative.py:152]: @coderabbitai `EDAData.__init__` converts all factor values to numpy arrays via `np.asarray()` at construction time. The values in `data.factors` are already `np.ndarray` objects — wrapping them again would be a no-op.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_comparative.py [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+
+    def test_legend_present(self, block_data):
+        _, ax = cp.block_plot(block_data)
+        assert ax.get_legend() is not None
+
+    def test_one_line_per_block(self, block_data):
+        _, ax = cp.block_plot(block_data)
+        n_blocks = len(np.unique(block_data.factors["block"]))
+        assert len(ax.get_lines()) >= n_blocks
+
```
[the remark this answers is not in the document]
Comment (Pjieter) [reply tests/test_comparative.py:97]: @coderabbitai `EDAData.__init__` validates that `y` is non-empty and that all `factors` arrays match `len(y)` at construction time. These edge cases would raise `ValueError` from `EDAData`, not from `block_plot` — so tests in this file would be testing `EDAData` validation, which is already covered in `tests/test_data.py`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_comparative.py [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+        _, ax_no_doe = cp.youden_plot(youden_data)
+        n_base = len(ax_no_doe.collections)
+        plt.close("all")
+        _, ax_doe = cp.youden_plot(youden_data, doe=True)
+        assert len(ax_doe.collections) == n_base + 1
+
+    def test_legend_present(self, youden_data):
+        _, ax = cp.youden_plot(youden_data)
+        assert ax.get_legend() is not None
+
```
[the remark this answers is not in the document]
Comment (Pjieter) [reply tests/test_comparative.py:140]: @coderabbitai `EDAData.__init__` validates that `y` is non-empty and that `x` matches `len(y)` at construction time. These edge cases would raise `ValueError` from `EDAData`, not from `youden_plot` — so tests here would be testing `EDAData` validation, which is already covered in `tests/test_data.py`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_comparative.py [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+        assert "A" in labels
+        assert "B" in labels
+        assert "C" in labels
+
+    def test_single_observation(self):
+        data = EDAData(y=[3.0], factors={"A": np.array([1.0])})
+        fig, ax = cp.star_plot(data)
+        assert isinstance(fig, Figure)
+        assert isinstance(ax, Axes)
+
```
[the remark this answers is not in the document]
Comment (Pjieter) [reply tests/test_comparative.py:180]: @coderabbitai `EDAData.__init__` validates that `y` is non-empty and that all `factors` arrays match `len(y)` at construction time. These edge cases would raise `ValueError` from `EDAData`, not from `star_plot` — so tests here would be testing `EDAData` validation, which is already covered in `tests/test_data.py`.

---
_Generated by [Claude Code](https://claude.ai/code)_
